storage/innobase/log/log0log.cc

```diff
@@ -1002,6 +1002,8 @@ template<bool release_latch> inline lsn_t log_t::write_buf() noexcept
         MEM_MAKE_DEFINED(re_write_buf + length, (write_size_1 + 1) - length);
 # endif
       buf[length]= 0; /* allow recovery to catch EOF faster */
+      if (UNIV_LIKELY_NULL(re_write_buf))
+        resize_buf[length]= 0;
```

Comment:
Instead of extending rather frequently executed low-level code, I think that we had better fix this bug in `log_t::write_checkpoint()`.

There is a small probability that `buf[length]` would already contain a 0 byte. That is not actually a problem for recovery. The `log_sys.buf` will always end at the end of a mini-transaction, that is, a byte 0 or 1 followed by a 4-byte checksum and a possible 8-byte encryption nonce. The log parser does not allow a 0 or 1 byte right after the end of a mini-transaction.

Reply:
I realized that changing this in `log_t::write_checkpoint()` would introduce a new system call and may require duplicating a lot of logic.

A better solution might be to replace the function `log_write_up_to()` with a function pointer, pointing to 3 or 4 different implementations:
 * one that invokes `log_t::persist()`, with or without log resizing being active
 * one that invokes `pwrite` to `ib_logfile0` while log resizing is not active
 * one that invokes `pwrite` to `ib_logfile0` and `ib_logfile101` while log resizing is active

The call in `log_write_and_flush()` should be infrequent and should not matter.

I will file a separate ticket for this refactoring.

Reply:
I ended up implementing this refactoring in 7ba1f87b5d128902e484f31852cfa6ac91ab1004. Surprisingly, the generated code for `log_writer<false>` would be larger than `log_writer<true>`, although I would expect it to be vice versa. I am investigating that.

Reply:
The explanation is that in `log_writer<true>`, the call to `log_t::write_buf<1>()` is not inline, while in the common case `log_writer<false>` it is. There is no non-`inline` instantiation of `template<int resizing> log_t::write_buf` with the template argument `0` in the executable that I checked, only with `-1` and `1`.

Reply:
I figured out that `__attribute__((always_inline))` does the trick: 9dd4ab5fc3fecf8da99ac40232ca8a8eef0b0037
For `log_t::resize_write_buf()` I will avoid forcing that, because the function is rather large (about 1/6 of the size of the caller) and would be invoked from two places: `log_t::write_buf<-1>` inlined in `log_writer<true>` and `log_t::write_buf<1>` inlined in `log_write_and_flush()`.